### PR TITLE
Refactor queue_options to use Pydantic dataclasses

### LINKED ISSUE #6712: Implement Pydantic model for queue options
We are already validating the strings in the config parser, but it would be nice to:

1. Cast values to their proper types (eg numbers as numbers, since we're in Python now they can be implicitly cast to strings when we compose the `bsub` command or whatever)
2. Apply default values in an obvious and maintainable way.
3. Replace the current list of tuples containing the options with a dataclass (say).

Then we can avoid having to do things like this in `JobQueue` in `queue.py`:

```python
def max_running(self) -> int:
    max_running = 0
    if self.driver.has_option("MAX_RUNNING"):
        max_running = int(self.driver.get_option("MAX_RUNNING"))
    if max_running == 0:
        return len(self._realizations)
    return max_running
```

### PR DESCRIPTION
**Pull request by berland**
**Issue**
Resolves #6712 

**Approach**
Rewrite

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/scheduler/__init__.py [hunk 1]
```diff
@@ -22,50 +24,39 @@ def create_driver(config: QueueConfig) -> Driver:
...
-            num_cpus_per_node=int(num_cpus_per_node) if num_cpus_per_node else None,
-            cluster_label=queue_options.get("CLUSTER_LABEL"),
-            job_prefix=queue_options.get("JOB_PREFIX"),
-            project_code=queue_options.get("PROJECT_CODE"),
-        )
+        queue_dict = asdict(config.queue_options)
+        queue_dict["queue_name"] = queue_dict.pop("queue")
+        queue_dict.pop("max_running")
+        queue_dict.pop("submit_sleep")
+        return OpenPBSDriver(**queue_dict)
```
Comment (berland) [inline src/ert/scheduler/__init__.py:31]: Should it be a helper function in the TorqueOptions class to provide this argument dictionary for all drivers?

File: src/ert/config/queue_config.py [hunk 2]
```diff
@@ -17,123 +19,167 @@
...
-        "PARTITION",
-        "INCLUDE_HOST",
-        "EXCLUDE_HOST",
-        "PROJECT_CODE",
-    ],
-    "TORQUE": [
-        "QSUB_CMD",
-        "QSTAT_CMD",
-        "QDEL_CMD",
-        "QSTAT_OPTIONS",
```
Comment (berland) [inline src/ert/config/queue_config.py:92]: This has been removed in the PR.

File: src/ert/config/queue_config.py [hunk 3]
```diff
@@ -17,123 +19,167 @@
...
-        "NUM_CPUS_PER_NODE",
-        "MAX_RUNNING",
-    ],
-    "LOCAL": ["MAX_RUNNING"],
-    "GENERIC": ["MAX_RUNNING"],
-}
-queue_positive_number_options: Mapping[str, List[str]] = {
-    "LSF": ["SUBMIT_SLEEP"],
-    "SLURM": ["SUBMIT_SLEEP", "SQUEUE_TIMEOUT", "MAX_RUNTIME"],
-    "TORQUE": ["SUBMIT_SLEEP", "QUEUE_QUERY_TIMEOUT"],
```
Comment (berland) [inline src/ert/config/queue_config.py:119]: queue_query_timeout is missing in PR

File: src/ert/scheduler/__init__.py (hunk 1 above)
Comment (berland) [reply src/ert/scheduler/__init__.py]: Now included as a bonus commit (to be squashed)

File: src/ert/config/queue_config.py (hunk 3 above)
Comment (berland) [reply src/ert/config/queue_config.py:119]: added

File: src/ert/config/queue_config.py (hunk 2 above)
Comment (berland) [reply src/ert/config/queue_config.py:92]: added.

File: src/ert/config/queue_config.py [hunk 4]
```diff
@@ -17,123 +19,210 @@
...
-    "MEMORY",
-    "MEMORY_PER_CPU",
-    "PARTITION",
-    "PROJECT_CODE",
-    "SBATCH",
-    "SCANCEL",
-    "SCONTROL",
-    "SQUEUE",
-    "SQUEUE_TIMEOUT",
+_all_queue_systems = [
```
Comment (berland) [inline src/ert/config/queue_config.py:22]: Unneeded, can just iterate over QueueSystem

File: src/ert/scheduler/slurm_driver.py [hunk 5]
```diff
@@ -73,8 +73,8 @@ def __init__(
         realization_memory: Optional[int] = 0,
         queue_name: Optional[str] = None,
         memory_per_cpu: Optional[str] = None,
-        max_runtime: Optional[str] = None,
-        squeue_timeout: Optional[str] = None,
+        max_runtime: Optional[float] = None,
+        squeue_timeout: float = 2,
```
Comment (xjules) [inline src/ert/scheduler/slurm_driver.py:77]: Is this value ever to be changed? ie. why to expose this to the users?

File: src/ert/config/queue_config.py [hunk 6]
```diff
@@ -17,123 +19,159 @@
...
+                    error["input"],
+                    is_selected_queue_system,
+                )
+            return None
+
+    def add_global_queue_options(self, config_dict: ConfigDict) -> None:
+        for generic_option in fields(QueueOptions):
+            if (
+                generic_value := config_dict.get(generic_option.name.upper(), None)  # type: ignore
+            ) and self.__dict__[generic_option.name] == generic_option.default:
```
Comment (xjules) [inline src/ert/config/queue_config.py:62]: what is this other constrain: `self.__dict__[generic_option.name] == generic_option.default:`?

File: src/ert/config/queue_config.py [hunk 7]
```diff
@@ -17,123 +19,159 @@
...
+    keep_qsub_output: bool = False
+
+    qstat_options: Optional[str] = pydantic.Field(default=None, deprecated=True)
+    queue_query_timeout: Optional[str] = pydantic.Field(default=None, deprecated=True)
+
+    @property
+    def driver_options(self) -> Dict[str, Any]:
+        driver_dict = asdict(self)
+        driver_dict["queue_name"] = driver_dict.pop("queue")
+        driver_dict.pop("max_running")
```
Comment (xjules) [inline src/ert/config/queue_config.py:127]: Why is `MAX_RUNNING` and `SUMBIT_SLEEP` removed?

File: src/ert/config/queue_config.py [hunk 8]
```diff
@@ -151,18 +189,68 @@ def validate(self, mem_str_format: str) -> bool:
...
+            )
+        processed_options[option_name] = value
+
+
+def _raise_for_defaulted_invalid_options(queue_config_list: List[List[str]]) -> None:
+    # Invalid options names with no values (i.e. defaulted) are not passed to
+    # the validation system, thus we neeed to catch them expliclitly
+    for queue_system, option_name, *_ in queue_config_list:
+        if option_name not in valid_options[queue_system]:
+            raise ConfigValidationError(
```
Comment (xjules) [inline src/ert/config/queue_config.py:226]: I guess this won't be an issue when running `LOCAL_QUEUE`; ie. when they won't remove settings for the sake of testing but they will change `QUEUE_SYSTEM` to local.

File: src/ert/scheduler/slurm_driver.py (hunk 5 above)
Comment (berland) [reply src/ert/scheduler/slurm_driver.py:77]: We keep it only for backwards compatibility.

File: src/ert/config/queue_config.py [hunk 9]
```diff
@@ -180,42 +268,82 @@ def from_dict(cls, config_dict: ConfigDict) -> QueueConfig:
...
-            selected_queue_system, queue_options, config_dict, realization_memory
-        )
+                queue_options.project_code = "+".join(tags)
+
+        if selected_queue_system == QueueSystem.TORQUE:
+            _check_num_cpu_requirement(
+                config_dict.get("NUM_CPU", 1), queue_options, _raw_queue_options
+            )
+
+        for _queue_vals in _all_validated_queue_options.values():
```
Comment (xjules) [inline src/ert/config/queue_config.py:316]: All these ifs can be solved by one `for loop` I guess :smile:

File: tests/unit_tests/config/test_queue_config.py [hunk 10]
```diff
@@ -276,18 +281,18 @@ def test_initializing_empty_config_queue_options_resets_to_default_value(
...
 
 
 @pytest.mark.usefixtures("use_tmpdir")
 @pytest.mark.parametrize(
     "queue_system, queue_option, queue_value, err_msg",
     [
-        ("SLURM", "SQUEUE_TIMEOUT", "5a", "is not a valid integer or float"),
-        ("TORQUE", "NUM_NODES", "3.5", "is not a valid positive integer"),
+        ("SLURM", "SQUEUE_TIMEOUT", "5a", "should be a valid number"),
+        ("TORQUE", "NUM_NODES", "3.5", "should be a valid integer"),
```
Comment (xjules) [inline tests/unit_tests/config/test_queue_config.py:295]: `should be a valid positive integer`

File: tests/unit_tests/config/test_queue_config.py [hunk 11]
```diff
@@ -524,22 +528,34 @@ def test_wrong_generic_queue_option_raises_validation_error(queue_system, key, v
...
     )
     with pytest.raises(ConfigValidationError, match=error_msg):
         ErtConfig.from_file("config.ert")
+
+
+@pytest.mark.parametrize(
+    "queue_system",
+    (QueueSystem.LSF, QueueSystem.TORQUE, QueueSystem.LOCAL, QueueSystem.SLURM),
+)
+def test_driver_initialization_from_defaults(queue_system):
```
Comment (xjules) [inline tests/unit_tests/config/test_queue_config.py:553]: should there be an assert or something in the end?

File: src/ert/config/queue_config.py [hunk 12]
```diff
@@ -18,123 +20,159 @@
...
+
+@pydantic.dataclasses.dataclass
+class LocalOptions(QueueOptions):
+    @property
+    def driver_options(self) -> Dict[str, Any]:
+        return {}
+
+
+@pydantic.dataclasses.dataclass
+class LsfOptions(QueueOptions):
```
Comment (xjules) [inline src/ert/config/queue_config.py:88]: Just thinking (aloud) here. Should we use the full names as class name; ie. `LsfQueueOptions`? Mainly `LocalOptions` sounds quite strange tbh.

File: src/ert/config/queue_config.py (hunk 6 above)
Comment (berland) [reply src/ert/config/queue_config.py:63]: It is to satisfy `test_global_config_key_does_not_overwrite_queue_options`

**Review by xjules — CHANGES_REQUESTED** [review]
Very nice job @berland ! Looks good, had a small comment on naming the `*Options` class and one test that requires some input. Otherwise :+1:

File: src/ert/config/queue_config.py (hunk 8 above)
Comment (berland) [reply src/ert/config/queue_config.py:227]: It is to fail hard whenever you have e.g. `QUEUE_OPTION LSF BLABLABLAOPTION foovalue` in your config.

File: src/ert/config/queue_config.py (hunk 9 above)
Comment (berland) [reply src/ert/config/queue_config.py:316]: from line 316 and onwards? Not so trivial I think?

File: tests/unit_tests/config/test_queue_config.py (hunk 11 above)
Comment (berland) [reply tests/unit_tests/config/test_queue_config.py:553]: The intention of the test was just to ensure that we can initialize without errors. If there are any spelling errors in the parameter names to the drivers init() this test will fail, also if we provide too many parameters. Testing that the arguments are processed correctly inside the drivers should be the responsibility of the drivers, not the config tests.

File: src/ert/config/queue_config.py (hunk 12 above)
Comment (berland) [reply src/ert/config/queue_config.py:88]: Good suggestion. Will fix.

File: src/ert/config/queue_config.py (hunk 9 above)
Comment (xjules) [reply src/ert/config/queue_config.py:316]: Yes, it is possible but not really pretty. Let's keep it.

**Review by xjules — APPROVED** [review]
Very good! :rocket:

**Pull request merged**